File: Session_06_DynamicProgramming_Classic/multi_language_implementations/longest_common_subsequence.cpp
#include <iostream>
#include <vector>
#include <string>
#include <algorithm>

void solveLCS(const std::string &s1, const std::string &s2) {
    int m = s1.length();
    int n = s2.length();
    std::vector<std::vector<int>> dp(m + 1, std::vector<int>(n + 1, 0));
    for (int i = 0; i <= m; i++) {
        for (int j = 0; j <= n; j++) {
            if (i == 0 || j == 0) dp[i][j] = 0;
            else if (s1[i - 1] == s2[j - 1]) dp[i][j] = dp[i - 1][j - 1] + 1;
            else dp[i][j] = std::max(dp[i - 1][j], dp[i][j - 1]);
        }
    }
    std::cout << "LCS Length: " << dp[m][n] << "\n";
    int index = dp[m][n];
    std::string lcsStr(index, ' ');
    int i = m, j = n;
    while (i > 0 && j > 0) {
        if (s1[i - 1] == s2[j - 1]) {
            lcsStr[index - 1] = s1[i - 1];
            i--; j--; index--;
        } else if (dp[i - 1][j] > dp[i][j - 1]) i--;
        else j--;
    }
    std::cout << "LCS String: " << lcsStr << "\n";
}

int main() {
    std::string s1 = "STONE";
    std::string s2 = "LONEST";
    solveLCS(s1, s2);
    std::cout << "\nPress Enter to exit...";
    std::cin.get();
    return 0;
}

File: Session_06_DynamicProgramming_Classic/multi_language_implementations/longest_common_subsequence.go
package main

import (
	"bufio"
	"fmt"
	"os"
)

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}

func solveLCS(s1 string, s2 string) {
	m := len(s1)
	n := len(s2)
	dp := make([][]int, m+1)
	for i := range dp {
		dp[i] = make([]int, n+1)
	}

	for i := 0; i <= m; i++ {
		for j := 0; j <= n; j++ {
			if i == 0 || j == 0 {
				dp[i][j] = 0
			} else if s1[i-1] == s2[j-1] {
				dp[i][j] = dp[i-1][j-1] + 1
			} else {
				dp[i][j] = max(dp[i-1][j], dp[i][j-1])
			}
		}
	}

	fmt.Printf("LCS Length: %d\n", dp[m][n])

	index := dp[m][n]
	lcsStr := make([]byte, index)
	i, j := m, n
	for i > 0 && j > 0 {
		if s1[i-1] == s2[j-1] {
			lcsStr[index-1] = s1[i-1]
			i--
			j--
			index--
		} else if dp[i-1][j] > dp[i][j-1] {
			i--
		} else {
			j--
		}
	}
	fmt.Printf("LCS String: %s\n", string(lcsStr))
}

func main() {
	s1 := "STONE"
	s2 := "LONEST"
	solveLCS(s1, s2)
	fmt.Printf("\nPress Enter to exit...")
	bufio.NewReader(os.Stdin).ReadBytes('\n')
}
